napari/plugins/utils.py

```diff
@@ -0,0 +1,67 @@
+from . import PluginError, log_plugin_error
+from . import plugin_manager as napari_plugin_manager
+from ._hookexec import _hookexec
+from typing import Optional
+from ..types import LayerData
+
+
+def get_layer_data_from_plugins(
+    path: str, plugin_manager=None
+) -> Optional[LayerData]:
+    """Iterate reader hooks and return first non-None LayerData or None.
+
+    This function returns as soon as the path has been read successfully,
+    while catching any plugin exceptions, storing them for later retrievial,
+    providing useful error messages, and relooping until either layer data is
+    returned, or no readers are found.
+
+    Exceptions will be caught and stored as PluginErrors
+    (in plugin_manager._exceptions)
+
+    Parameters
+    ----------
+    path : str
+        The path (file, directory, url) to open
+    plugin_manager : pluggy.PluginManager, optional
+        Instance of a pluggy PluginManager.  by default the main napari
+        plugin_manager will be used.
+
+    Returns
+    -------
+    LayerData or None
+        LayerData that can be *passed to _add_layer_from_data.  If no reader
+        plugins are (or they all error), returns None
+    """
+    plugin_manager = plugin_manager or napari_plugin_manager
+    skip_imps = []
+    while True:
+        (reader, imp) = _hookexec(
```

Comment:
`(reader, implementation) = `?

Reply:
I'd kind of like a comment above this about what hookexec does. It seems very magical.

Reply:
I will add a comment to the code, but here's the explanation... buckle up ;)

When using a pluggy `PluginManager`, when you actually call a hook using their syntax `plugin_manager.hook.hook_spec_name(**kwargs)` it goes through a relatively complicated chain of methods and aliases before ultimately landing on the basic function that loops through plugins:

[`hooks._HookCaller.__call__`](https://github.com/pytest-dev/pluggy/blob/6e8b6d4e8e0519e50af05b4e4b90b18b924bcbcc/src/pluggy/hooks.py#L278) → 
[`hooks._HookCaller._hookexec`](https://github.com/pytest-dev/pluggy/blob/6e8b6d4e8e0519e50af05b4e4b90b18b924bcbcc/src/pluggy/hooks.py#L292) → created during hook registration in
[`PluginManager.register`](https://github.com/pytest-dev/pluggy/blob/6e8b6d4e8e0519e50af05b4e4b90b18b924bcbcc/src/pluggy/manager.py#L124) →
[`PluginManager._hookexec`](https://github.com/pytest-dev/pluggy/blob/6e8b6d4e8e0519e50af05b4e4b90b18b924bcbcc/src/pluggy/manager.py#L90)  →
[`PluginManager._inner_hookexec`](https://github.com/pytest-dev/pluggy/blob/6e8b6d4e8e0519e50af05b4e4b90b18b924bcbcc/src/pluggy/manager.py#L84) → which really just calls
[`_HookCaller.multicall`](https://github.com/pytest-dev/pluggy/blob/6e8b6d4e8e0519e50af05b4e4b90b18b924bcbcc/src/pluggy/hooks.py#L215) → which is an alias for
[`pluggy.callers._multicall`](https://github.com/pytest-dev/pluggy/blob/6e8b6d4e8e0519e50af05b4e4b90b18b924bcbcc/src/pluggy/callers.py#L157) (_this is the the thing that actually loops through plugin hook_implementations_)

**point being**: if you want to make any modifications to the way plugins are looped through in a `PluginManager`, one can't really just override a method on our `PluginManager` class.  My `_hookexec` function here is my strategy for modifying the core plugin loop.  I will update the comments/name of that function to make that clearer, since I suspect it may be a permanent resident in napari (given the importance of that plugin loop).